Add option to enable Super Seeding mode once ratio/time limit is reached. Closes #7160.

thalieht opened the pull request:
When ratio/time limit is reached, ETA shows zero for affected torrents until qbt restart, after that, infinity. Thought i should point it out in case it's undesirable.

evsh commented on src/gui/optionsdlg.ui:
```diff
@@ -122,8 +122,8 @@
              <y>0</y>
-             <width>501</width>
-             <height>893</height>
+             <width>560</width>
```
Are all of these size changes required?

thalieht replied:
Qt creator in design mode did them automatically when it saved the file, any idea how to stop it?

evsh replied:
No, sorry. I use standalone Designer app, and it does not resize forms on opening.

thalieht replied:
I'll try that.

thalieht replied:
Still the same.

evsh replied:
It may happen when your visual style requires widgets to be bigger than that used be the previous author and designer app has to expand the form to accommodate expanded widgets.

thalieht replied:
Yeah probably, i reset my system font sizes to default and although it still changes these values, they are closer to the original.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -1796,10 +1800,14 @@ void Session::processShareLimits()
                             deleteTorrent(torrent->hash());
                         }
-                        else if (!torrent->isPaused()) {
+                        else if ((m_maxRatioAction == Pause) && !torrent->isPaused() ) {
```
Redundant space after isPaused()

glassez replied:
This method has the same code in 2 places (for max ratio and for max seeding time). Maybe to rearrange it?

thalieht replied:
>Maybe to rearrange it?

I'd rather not.

glassez commented on src/base/bittorrent/session.h:
```diff
@@ -121,7 +121,8 @@ class ResumeDataSavingManager;
     Pause,
-    Remove
+    Remove,
+    SuperSeeding
```
EnableSuperSeeding

glassez commented on src/gui/optionsdlg.ui:
```diff
@@ -2420,6 +2420,11 @@
                  </item>
+                 <item>
+                  <property name="text">
+                   <string>Super seed them</string>
```
Is it correct wording?

thalieht replied:
I'm not sure. At first i had " Enable super seeding for them" but it seems excessively long.

thalieht replied:
@LordNyriox any ideas?

LordNyriox replied:
@thalieht:  I do not even understand what `them` refers to.  

A pronoun without an antecedent is confusing as heck.  

That being said, *if* `them` refers to torrents I would write `Use super seeding mode for them`—since `super seeding` in this case is a mode, rather than an action.

LordNyriox replied:
@thalieht:  `Enable super seeding for them` works as well—since using the word "enable" implies a binary state (AKA, a "mode").

LordNyriox replied:
@thalieht:  `Super seed them`, on the other hand, does NOT work—since it implies that the actual action being performed is "super seeding" itself, rather than "enabling super seeding mode".  

English is fraught with pitfalls like that.

thalieht replied:
Sorry, indeed 'them' refers to torrents. Thanks! Guess i'll go with "Enable super seeding for them" unless there are further objections.

thalieht commented:
I separated the changes in the .ui file in their own commit until we sort this out.
Here is what i did in Designer:

1. Break the grid layout of `ratioBox`
2. Deleted the spacer left of "then" as it was causing this minor issue:
![seperating2](https://user-images.githubusercontent.com/6451685/30733968-f44b402a-9f81-11e7-8f9a-ac4a3ec9d123.gif)
3. Made a horizontal layout with `comboRatioLimitAct` and the spacer on it's right otherwise it would elide "Enable super seeding mode" in the combobox.
4. Added colons (:) at the end of checkboxes's strings as i noticed they are the only ones without them (should i add it to "then" as well?).
5. Added the "Enable super seeding mode" in `comboRatioLimitAct`.
6. Re-set layout of `ratioBox` back to grid.

And now it's pretty much the same (don't compare with the .gif as it's from beta2 in Windows)
![capture](https://user-images.githubusercontent.com/6451685/30734463-ed33c364-9f83-11e7-97e2-4455d9a02dab.JPG)

thalieht commented:
Is there any point to this PR? I mean, if i'm not mistaken, super seeding is to help seed a new or dying torrent with few seeds and many peers. What does that have to do with the personal up/down ratio and/or seeding time?

Personally i don't care about this feature so i'm fine with whatever you decide, i just saw the relevant issue and thought i might be able to help with it.

LordNyriox commented:
@thalieht:  

>Is there any point to this PR? 

I admit that I was wondering that myself… 

>I mean, if i'm not mistaken, super seeding is to help seed a new or dying torrent with few seeds and many peers. 

You are exactly correct.  

>What does that have to do with the personal up/down ratio and/or seeding time? 

Nothing at all—or at least, nothing relevant to the usage of a torrent.  

What I really want to know is, why such a niche issue was not rejected outright?

thalieht commented:
As i thought... thanks @LordNyriox for the answer. Let's see if someone thinks differently.

LordNyriox commented:
@thalieht:  Perhaps you could move the **fixes** for coding style and UI into a separate PR?  

That way they can be merged, even if this PR is closed.

thalieht commented:
The only fix here is the deletion of the spacer in the .ui file, but considering i'm not even sure this is the correct approach and that the only people reviewing .ui changes are sledgehammer and chocobo1, sledgehammer doesn't seem very active lately and judging from the last commit message in webui wiki, i don't think chocobo1 is too keen on "babysitting" me.
So from now on i'll just avoid changing .ui files.

LordNyriox commented:
@thalieht:  

>The only fix here is the deletion of the spacer in the .ui file 

I thought there were coding style fixes as well…

thalieht commented:
Nope.

thalieht commented:
Closing because this feature is probably a bad idea. Someone close #7160

KaWiSo commented:
Why is the option to enable super seeding for torrents that reached the ratio considered a bad idea?
It would allow you to manage your bandwidth, while still keeping all torrents alive.

thalieht commented:
>super seeding is to help seed a new or dying torrent with few seeds and many peers. What does that have to do with the personal up/down ratio and/or seeding time?

thalieht commented:
Re-opening because of https://github.com/qbittorrent/qBittorrent/issues/7160#issuecomment-408357826
Rebased and reworked the .ui see the updated https://github.com/qbittorrent/qBittorrent/pull/7475#issuecomment-331378223 for what i did and ignore the pics in that post.

Before:
![before](https://user-images.githubusercontent.com/6451685/43331632-f6a39990-91ce-11e8-9177-ecd68f7cb594.JPG)

After:
![after](https://user-images.githubusercontent.com/6451685/43331649-fd3611a2-91ce-11e8-896f-d86895bd741f.JPG)

thalieht commented:
Rebased. Also added the `&& isPaused()` condition to enable it because it was being enabled for paused/completed torrents.
I dropped all changes in the .ui file except the align text to the left in neighboring spinboxes. The problem demonstrated in https://github.com/qbittorrent/qBittorrent/pull/7475#issuecomment-331378223 .gif no longer happens in master.

GUI
![Capture5](https://user-images.githubusercontent.com/6451685/62724421-9d73b500-ba1b-11e9-89a0-9a69eca4b4fe.JPG)
WebUI
![Capture6](https://user-images.githubusercontent.com/6451685/62724435-a3699600-ba1b-11e9-9ff9-34ddf61ff8d0.JPG)

glassez commented:
>because it was being enabled for paused/completed torrents.

How can paused torrent reach seeding limit?

thalieht commented:
Ratio/seeding time limit has been reached, while not paused obviously, and then user decides to change to "enable super seeding for them". If the torrent has fulfilled the criteria it gets super seeding enabled, paused or otherwise.

Chocobo1 commented on src/gui/optionsdialog.ui:
```diff
@@ -2456,7 +2456,7 @@
                  </property>
                  <property name="alignment">
-                  <set>Qt::AlignCenter</set>
+                  <set>Qt::AlignLeading|Qt::AlignLeft|Qt::AlignVCenter</set>
```
if you only want left alignment then just remove the whole `<property name="alignment>` section.
i.e. remove the following:
```xml
<property name="alignment">
  <set>Qt::AlignCenter</set>
</property>
```

thalieht replied:
Done.

Chocobo1 commented on src/gui/optionsdialog.ui:
```diff
@@ -2516,7 +2521,7 @@
                  </property>
                  <property name="alignment">
-                  <set>Qt::AlignHCenter</set>
+                  <set>Qt::AlignLeading|Qt::AlignLeft|Qt::AlignVCenter</set>
```
same here.

glassez commented:
>Ratio/seeding time limit has been reached, while not paused obviously, and then user decides to change to "enable super seeding for them". If the torrent has fulfilled the criteria it gets super seeding enabled, paused or otherwise.

It seems that the current algorithm is not smart enough to understand whether the limit was reached now or sometime earlier.

thalieht commented:
>It seems that the current algorithm is not smart enough to understand whether the limit was reached now or sometime earlier.

¯\\_(ツ)_/¯

Chocobo1 commented:
@thalieht 
Thank you!
